12-seongwon030

seongwon030 opened the pull request:
## 🔗 문제 링크
[암호 만들기](https://www.acmicpc.net/problem/1759)

## ✔️ 소요된 시간
2시간

## ✨ 수도 코드
브루트포스라고 하니 모든 경우의 수를 찾는 게 우선이라고 생각했다. 처음엔 <code>모음리스트</code>와 <code>자음리스트</code>로 나누어 모음리스트 조합과 자음리스트 조합을 따로 만들어 두 개를 합치는 방식으로 했더니 출력초과가 났다. 다른 테스트케이스를 다 넣어봤는데도 통과되지 않았다..

### 힌트
질문게시판에서 테스트케이스를 찾아 보다가 어떤 사람의 답변이 힌트가 되었다. L,C가 15이하면 모든 조합의 개수가 C중 L개를 뽑는 경우의 수인데 그렇게 많은 경우의 수가 필요하지 않겠다는 생각이 들었다. 그래서 <code>모든 경우의 조합을 순회</code>하면서 그 조합에 <code>모음이 최소 1개 그리고 자음이 최소 2개 있는지 체크</code>하면서 조건에 만족하면 <code>result</code> 배열에 넣은 다음 마지막에 sort()를 써서 출력했다. 

### 주의할 것
```python
  vowel_str = ''.join(sorted(c))
```
조합 문자열을 체크할 때 <code>sorted()</code>로 문자열을 정렬해줘야 나중에 사전식으로 출력이 된다. 

## 📚 새롭게 알게된 내용
브루트포스는 무식하게 하는게 맞는 것 같다. 일단 모든 경우의 수를 생각해보고 안 되면 다음 방법으로 가는 방향으로..

InSange reviewed (approved):
오름차순이고 크기가 15일때 아무리 커도 8자리 * (26 알파벳에서 절반의 가능성 b-> (j < alpha)) 순으로 될 것 같습니다. 그럼 올 수 있는 가능성은 8!정도? 될 수 있겠네요..
그럼 아마 나올수 있는 오름차순 단어의 조합은 322,560정도가 나올 것으로 예상할 수 있다고 봅니다.

저는 먼저 sort를 해서 맨 앞 요소부터 L 크기까지 비교를 진행할 것 같습니다.
모음 요소를 비교해줄 함수를 하나 구현하여 첫번째 단어 부터 L 크기까지 단어를 추가한 string에서 만약 함수에서 모음이 포함되어 있다고 판별될 시 bool형 flag 변수를 통해 해당 단어는 조건에 만족한다고 표시하여 벡터에 넣는 식으로 진행한다면 어떨까요?

seongwon030 commented:
> 오름차순이고 크기가 15일때 아무리 커도 8자리 * (26 알파벳에서 절반의 가능성 b-> (j < alpha)) 순으로 될 것 같습니다. 그럼 올 수 있는 가능성은 8!정도? 될 수 있겠네요.. 그럼 아마 나올수 있는 오름차순 단어의 조합은 322,560정도가 나올 것으로 예상할 수 있다고 봅니다.
> 
> 저는 먼저 sort를 해서 맨 앞 요소부터 L 크기까지 비교를 진행할 것 같습니다. 모음 요소를 비교해줄 함수를 하나 구현하여 첫번째 단어 부터 L 크기까지 단어를 추가한 string에서 만약 함수에서 모음이 포함되어 있다고 판별될 시 bool형 flag 변수를 통해 해당 단어는 조건에 만족한다고 표시하여 벡터에 넣는 식으로 진행한다면 어떨까요?

```python
from itertools import combinations

L,C = map(int,input().split())

k = list(map(str,input().split()))
k.sort()

vowels = 'aeiou'

com = combinations(k,L)

for c in com:
  vowel_cnt = sum(1 for char in c if char in vowels)
  consonant_cnt = L - vowel_cnt
  if vowel_cnt>=1 and consonant_cnt>=2:
    print(''.join(c))
```
그렇게 짜봤는데요. 모음의 개수가 1개 이상인지만 판별하게 되면 L=3일 때 모음의 개수가 2개면 자음의 1개 존재하는 예외상황처럼 자음이 최소 2개 존재한다는 조건을 간과하게 되는 것 같아요. 그래도 먼저 sort()하고 코드를 짜니까 더 깔끔해진 것 같습니다.

yuyu0830 commented on seongwon030/브루트포스/암호만들기.py:
```diff
@@ -0,0 +1,21 @@
+  vowel_str = ''.join(sorted(c))
+  vowel_cnt = sum(1 for char in vowel_str if char in vowels)
+  consonant_cnt = L - vowel_cnt
+  if vowel_cnt >=1 and consonant_cnt >=2:
```
간단한 팁인데 `x == 1' 이나 `x >= 1` 같은 코드는 그냥 `x`로 쓰면 같은 효과를 얻을 수 있습니다!

yuyu0830 reviewed (approved):
미리 정렬하고 탐색했다면 조금 더 시간을 줄일 수 있었을 것 같네요..! 특히 미리 정렬을 해뒀다면 정렬 순서가 보장되기 때문에 append로 결과를 모을 필요 없이 그때그때 출력해도 된다는 장점이 있습니다!

파이썬 스타일로 모음 개수 세는 파트가 신기하네요 ㅋㅋ

dhlee777 reviewed (approved):
처음부터` 모음리스트와 자음리스트`를 나누는것 보다 `모든경우의 수`를 탐색해주며 조합에서 모음이 최소 1개 그리고 자음이 최소 2개인경우에만 `result `에 넣어주는 브루트포스 방식이네요  이렇게 `경우의수`가 많지않은경우 브루트포스같은 노가다도  효율적인 것 같습니다.
